File: 3962-maximum-subarray-sum-after-at-most-k-swaps/3962-maximum-subarray-sum-after-at-most-k-swaps.php
class Solution {

    /**
     * @param Integer[] $nums
     * @param Integer $k
     * @return Integer
     */
    function maxSum($nums, $k) {

        $n = count($nums);

        $ans = PHP_INT_MIN;

        if ($k == 0 || $n == 1) {
            $sm = 0;

            foreach ($nums as $num) {
                $sm += $num;
                $ans = max($ans, $sm);

                if ($sm < 0) {
                    $sm = 0;
                }
            }

            return $ans;
        }

        $cnt = 0;
        $cur = 0;
        $pref = array_fill(0, $n + 1, 0);

        for ($i = 0; $i < $n; $i++) {
            if ($nums[$i] >= 0) {
                $cur += $nums[$i];
            } else {
                $cnt++;
            }

            $pref[$i + 1] = $pref[$i] + $nums[$i];
            $ans = max($ans, $nums[$i]);
        }

        if ($cnt <= $k) {
            return $cur;
        }

        $dp = array_fill(0, $n, array_fill(0, $n, 0));

        for ($i = 0; $i < $n; $i++) {
            $pq = [];
            $sm = 0;

            for ($j = $i; $j < $n; $j++) {
                if ($nums[$j] >= 0) {
                    $dp[$i][$j] = $sm;
                    continue;
                }

                if (count($pq) < $k) {
                    $pq[] = $nums[$j];
                    rsort($pq);
                    $sm += $nums[$j];
                } else {
                    if ($pq[0] > $nums[$j]) {
                        $sm -= $pq[0];
                        $pq[0] = $nums[$j];
                        rsort($pq);
                        $sm += $nums[$j];
                    }
                }
            }
        }

        for ($i = 0; $i < $n; $i++) {
            $pqmax = [];
            $sm = 0;

            for ($j = 0; $j < $i; $j++) {
                if ($nums[$j] < 0) {
                    continue;
                }

                if (count($pqmax) < $k) {
                    $pqmax[] = $nums[$j];
                    sort($pqmax);
                    $sm += $nums[$j];
                } else {
                    if ($pqmax[0] < $nums[$j]) {
                        $sm -= $pqmax[0];
                        $pqmax[0] = $nums[$j];
                        sort($pqmax);
                        $sm += $nums[$j];
                    }
                }
            }

            for ($j = $n - 1; $j > $i; $j--) {
                $curSum = $pref[$j + 1] - $pref[$i];
                $curSum -= $dp[$i][$j];
                $curSum += $sm;

                $ans = max($ans, $curSum);

                if ($nums[$j] < 0) {
                    continue;
                }

                if (count($pqmax) < $k) {
                    $pqmax[] = $nums[$j];
                    sort($pqmax);
                    $sm += $nums[$j];
                } else {
                    if ($pqmax[0] < $nums[$j]) {
                        $sm -= $pqmax[0];
                        $pqmax[0] = $nums[$j];
                        sort($pqmax);
                        $sm += $nums[$j];
                    }
                }

                if (!empty($pqmax)) {
                    $ans = max($ans, $sm);
                }
            }
        }

        return $ans;
    }
}

File: 3962-maximum-subarray-sum-after-at-most-k-swaps/3962-maximum-subarray-sum-after-at-most-k-swaps.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
def max_sum(nums, k)
  n = nums.size

  ans = -(1 << 31)

  if k == 0 || n == 1
    sm = 0

    nums.each do |num|
      sm += num
      ans = [ans, sm].max
      sm = 0 if sm < 0
    end

    return ans
  end

  cnt = 0
  cur = 0
  pref = Array.new(n + 1, 0)

  n.times do |i|
    if nums[i] >= 0
      cur += nums[i]
    else
      cnt += 1
    end

    pref[i + 1] = pref[i] + nums[i]
    ans = [ans, nums[i]].max
  end

  return cur if cnt <= k

  dp = Array.new(n) { Array.new(n, 0) }

  n.times do |i|
    pq = []
    sm = 0

    (i...n).each do |j|
      if nums[j] >= 0
        dp[i][j] = sm
        next
      end

      if pq.size < k
        pq << nums[j]
        pq.sort!.reverse!
        sm += nums[j]
      elsif pq[0] > nums[j]
        sm -= pq[0]
        pq[0] = nums[j]
        pq.sort!.reverse!
        sm += nums[j]
      end
    end
  end

  n.times do |i|
    pqmax = []
    sm = 0

    (0...i).each do |j|
      next if nums[j] < 0

      if pqmax.size < k
        pqmax << nums[j]
        pqmax.sort!
        sm += nums[j]
      elsif pqmax[0] < nums[j]
        sm -= pqmax[0]
        pqmax[0] = nums[j]
        pqmax.sort!
        sm += nums[j]
      end
    end

    (n - 1).downto(i + 1) do |j|
      cur_sum = pref[j + 1] - pref[i]
      cur_sum -= dp[i][j]
      cur_sum += sm

      ans = [ans, cur_sum].max

      next if nums[j] < 0

      if pqmax.size < k
        pqmax << nums[j]
        pqmax.sort!
        sm += nums[j]
      elsif pqmax[0] < nums[j]
        sm -= pqmax[0]
        pqmax[0] = nums[j]
        pqmax.sort!
        sm += nums[j]
      end

      ans = [ans, sm].max unless pqmax.empty?
    end
  end

  ans
end